src/planner/rules/expr.rs

```diff
@@ -172,6 +184,47 @@ fn is_not_zero(var: &str) -> impl Fn(&mut EGraph, Id, &Subst) -> bool {
     value_is(var, |v| !v.is_zero())
 }
 
+fn is_greater_than_or_equal(var1: &str, var2: &str) -> impl Fn(&mut EGraph, Id, &Subst) -> bool {
+    value_cmp(var1, var2, |d1, d2| {
+        std::mem::discriminant(d1) == std::mem::discriminant(d2) && d1.ge(d2)
+    })
+}
+
+fn is_greater_than(var1: &str, var2: &str) -> impl Fn(&mut EGraph, Id, &Subst) -> bool {
+    value_cmp(var1, var2, |d1, d2| {
+        std::mem::discriminant(d1) == std::mem::discriminant(d2) && d1.gt(d2)
+    })
+}
+
+fn is_less_than_or_equal(var1: &str, var2: &str) -> impl Fn(&mut EGraph, Id, &Subst) -> bool {
+    value_cmp(var1, var2, |d1, d2| {
+        std::mem::discriminant(d1) == std::mem::discriminant(d2) && d1.le(d2)
+    })
+}
+
+fn is_less_than(var1: &str, var2: &str) -> impl Fn(&mut EGraph, Id, &Subst) -> bool {
+    value_cmp(var1, var2, |d1, d2| {
+        std::mem::discriminant(d1) == std::mem::discriminant(d2) && d1.lt(d2)
+    })
+}
+
+fn value_cmp(
+    v1: &str,
+    v2: &str,
+    f: impl Fn(&DataValue, &DataValue) -> bool,
+) -> impl Fn(&mut EGraph, Id, &Subst) -> bool {
+    let v1 = var(v1);
+    let v2 = var(v2);
+
+    move |egraph, _, subst| match (
+        &egraph[subst[v1]].data.constant,
+        &egraph[subst[v2]].data.constant,
+    ) {
+        (Some(d1), Some(d2)) => f(d1, d2),
```

Comment:
Seems like the common condition `std::mem::discriminant(d1) == std::mem::discriminant(d2)` can be moved to here.